File: backend/internal/handlers/categories.go
package handlers

import (
	"encoding/json"
	"net/http"

	"github.com/jackc/pgx/v5/pgtype"
	"github.com/joselitophala/budget-planner-backend/internal/auth"
	"github.com/joselitophala/budget-planner-backend/internal/models"
	"github.com/joselitophala/budget-planner-backend/internal/utils"
)

// CategoryHandler handles category-related requests
type CategoryHandler struct {
	queries *models.Queries
}

// NewCategoryHandler creates a new category handler
func NewCategoryHandler(queries *models.Queries) *CategoryHandler {
	return &CategoryHandler{queries: queries}
}

// CategoryResponse represents a category in API responses
type CategoryResponse struct {
	ID           string   `json:"id"`
	Name         string   `json:"name"`
	Icon         *string  `json:"icon,omitempty"`
	Color        string   `json:"color"`
	IsSystem     bool     `json:"isSystem"`
	DefaultLimit *float64 `json:"defaultLimit,omitempty"`
}

// CreateCategoryRequest represents the create category request
type CreateCategoryRequest struct {
	Name         string  `json:"name"`
	Icon         *string `json:"icon,omitempty"`
	Color        string  `json:"color"`
	DefaultLimit *float64 `json:"defaultLimit,omitempty"`
}

// UpdateCategoryRequest represents the update category request
type UpdateCategoryRequest struct {
	Name         *string `json:"name,omitempty"`
	Icon         *string `json:"icon,omitempty"`
	Color        *string `json:"color,omitempty"`
	DefaultLimit *float64 `json:"defaultLimit,omitempty"`
}

// ListCategories returns all categories for the current user
func (h *CategoryHandler) ListCategories(w http.ResponseWriter, r *http.Request) {
	userID, ok := auth.GetUserID(r)
	if !ok {
		utils.Unauthorized(w, "Not authenticated")
		return
	}

	categories, err := h.queries.GetUserCategories(r.Context(), utils.PgUUID(userID))
	if err != nil {
		utils.InternalError(w, "Failed to fetch categories")
		return
	}

	response := make([]CategoryResponse, len(categories))
	for i, cat := range categories {
		response[i] = CategoryResponse{
			ID:           cat.ID,
			Name:         cat.Name,
			Icon:         utils.TextToStringPtr(cat.Icon),
			Color:        utils.TextToString(cat.Color),
			IsSystem:     cat.IsSystem.Valid,
			DefaultLimit: utils.NumericToFloat64Ptr(cat.DefaultLimit),
		}
	}

	utils.SendSuccess(w, response)
}

// GetSystemCategories returns all system default categories
func (h *CategoryHandler) GetSystemCategories(w http.ResponseWriter, r *http.Request) {
	categories, err := h.queries.GetSystemCategories(r.Context())
	if err != nil {
		utils.InternalError(w, "Failed to fetch system categories")
		return
	}

	response := make([]CategoryResponse, len(categories))
	for i, cat := range categories {
		response[i] = CategoryResponse{
			ID:           cat.ID,
			Name:         cat.Name,
			Icon:         utils.TextToStringPtr(cat.Icon),
			Color:        utils.TextToString(cat.Color),
			IsSystem:     cat.IsSystem.Valid,
			DefaultLimit: utils.NumericToFloat64Ptr(cat.DefaultLimit),
		}
	}

	utils.SendSuccess(w, response)
}

// CreateCategory creates a new custom category
func (h *CategoryHandler) CreateCategory(w http.ResponseWriter, r *http.Request) {
	userID, ok := auth.GetUserID(r)
	if !ok {
		utils.Unauthorized(w, "Not authenticated")
		return
	}

	var req CreateCategoryRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.BadRequest(w, "Invalid request body")
		return
	}

	var icon pgtype.Text
	if req.Icon != nil {
		icon = utils.PgText(*req.Icon)
	}

	var defaultLimit pgtype.Numeric
	if req.DefaultLimit != nil {
		defaultLimit = utils.PgNumeric(*req.DefaultLimit)
	}

	category, err := h.queries.CreateCategory(r.Context(), models.CreateCategoryParams{
		UserID:       utils.PgUUID(userID),
		Name:         req.Name,
		Icon:         icon,
		Color:        utils.PgText(req.Color),
		IsSystem:     pgtype.Bool{Valid: true, Bool: false},
		DefaultLimit: defaultLimit,
	})
	if err != nil {
		utils.InternalError(w, "Failed to create category")
		return
	}

	utils.SendCreated(w, CategoryResponse{
		ID:           category.ID,
		Name:         category.Name,
		Icon:         utils.TextToStringPtr(category.Icon),
		Color:        utils.TextToString(category.Color),
		IsSystem:     category.IsSystem.Valid,
		DefaultLimit: utils.NumericToFloat64Ptr(category.DefaultLimit),
	})
}

// UpdateCategory updates an existing category
func (h *CategoryHandler) UpdateCategory(w http.ResponseWriter, r *http.Request) {
	_, ok := auth.GetUserID(r)
	if !ok {
		utils.Unauthorized(w, "Not authenticated")
		return
	}

	categoryID := r.PathValue("id")
	if categoryID == "" {
		utils.BadRequest(w, "Category ID is required")
		return
	}

	var req UpdateCategoryRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.BadRequest(w, "Invalid request body")
		return
	}

	var name pgtype.Text
	if req.Name != nil {
		name = utils.PgText(*req.Name)
	}

	var icon pgtype.Text
	if req.Icon != nil {
		icon = utils.PgText(*req.Icon)
	}

	var color pgtype.Text
	if req.Color != nil {
		color = utils.PgText(*req.Color)
	}

	var defaultLimit pgtype.Numeric
	if req.DefaultLimit != nil {
		defaultLimit = utils.PgNumeric(*req.DefaultLimit)
	}

	category, err := h.queries.UpdateCategory(r.Context(), models.UpdateCategoryParams{
		ID:           categoryID,
		Name:         name,
		Icon:         icon,
		Color:        color,
		DefaultLimit: defaultLimit,
	})
	if err != nil {
		utils.InternalError(w, "Failed to update category")
		return
	}

	utils.SendSuccess(w, CategoryResponse{
		ID:           category.ID,
		Name:         category.Name,
		Icon:         utils.TextToStringPtr(category.Icon),
		Color:        utils.TextToString(category.Color),
		IsSystem:     category.IsSystem.Valid,
		DefaultLimit: utils.NumericToFloat64Ptr(category.DefaultLimit),
	})
}

// DeleteCategory soft deletes a category
func (h *CategoryHandler) DeleteCategory(w http.ResponseWriter, r *http.Request) {
	_, ok := auth.GetUserID(r)
	if !ok {
		utils.Unauthorized(w, "Not authenticated")
		return
	}

	categoryID := r.PathValue("id")
	if categoryID == "" {
		utils.BadRequest(w, "Category ID is required")
		return
	}

	err := h.queries.DeleteCategory(r.Context(), categoryID)
	if err != nil {
		utils.InternalError(w, "Failed to delete category")
		return
	}

	utils.SendSuccess(w, map[string]string{
		"message": "Category deleted successfully",
	})
}

File: frontend/src/lib/api/categories.ts
import { apiClient } from './client';
import type { Category } from '$lib/db/schema';

/**
 * Category API types (matching backend responses)
 */

export interface CategoryResponse {
	id: string;
	name: string;
	icon?: string;
	color: string;
	isSystem: boolean;
	defaultLimit?: number;
}

export interface CreateCategoryRequest {
	name: string;
	icon?: string;
	color: string;
	defaultLimit?: number;
}

export interface UpdateCategoryRequest {
	name?: string;
	icon?: string;
	color?: string;
	defaultLimit?: number;
}

/**
 * Type adapters to convert between backend and frontend types
 */

/**
 * Convert frontend Category to backend CreateCategoryRequest
 */
export function adaptCategoryToBackend(category: Category): CreateCategoryRequest {
	return {
		name: category.name,
		icon: category.icon || undefined,
		color: category.color || '#000000',
		defaultLimit: category.defaultLimit || undefined
	};
}

/**
 * Convert backend CategoryResponse to frontend Category
 */
export function adaptCategoryToFrontend(response: CategoryResponse, userId: string | null): Category {
	return {
		id: response.id,
		userId, // null for system categories
		name: response.name,
		icon: response.icon || null,
		color: response.color,
		isDefault: response.isSystem,
		defaultLimit: response.defaultLimit || null,
		createdAt: new Date().toISOString(), // Backend doesn't return this
		updatedAt: new Date().toISOString()  // Backend doesn't return this
	};
}

/**
 * Categories API client
 */
export class CategoriesApi {
	/**
	 * Get all categories for the current user
	 * GET /api/categories
	 */
	async getAllCategories(userId: string): Promise<Category[]> {
		const response = await apiClient.get<CategoryResponse[]>('/categories');
		return response.map(cat => adaptCategoryToFrontend(cat, userId));
	}

	/**
	 * Get system categories
	 * GET /api/categories/system
	 */
	async getSystemCategories(): Promise<Category[]> {
		const response = await apiClient.get<CategoryResponse[]>('/categories/system');
		return response.map(cat => adaptCategoryToFrontend(cat, null)); // System categories have null userId
	}

	/**
	 * Get a single category by ID
	 * GET /api/categories/{id}
	 */
	async getCategoryById(id: string, userId: string): Promise<Category> {
		const response = await apiClient.get<CategoryResponse>(`/categories/${id}`);
		return adaptCategoryToFrontend(response, userId);
	}

	/**
	 * Create a new category
	 * POST /api/categories
	 */
	async createCategory(category: Category): Promise<Category> {
		const request = adaptCategoryToBackend(category);
		const response = await apiClient.post<CategoryResponse>('/categories', request);
		return adaptCategoryToFrontend(response, category.userId);
	}

	/**
	 * Update a category
	 * PUT /api/categories/{id}
	 */
	async updateCategory(id: string, category: Category): Promise<Category> {
		const request: UpdateCategoryRequest = {
			name: category.name,
			icon: category.icon || undefined,
			color: category.color || undefined,
			defaultLimit: category.defaultLimit || undefined
		};

		const response = await apiClient.put<CategoryResponse>(`/categories/${id}`, request);
		return adaptCategoryToFrontend(response, category.userId);
	}

	/**
	 * Delete a category
	 * DELETE /api/categories/{id}
	 */
	async deleteCategory(id: string): Promise<void> {
		return apiClient.delete<void>(`/categories/${id}`);
	}
}

// Export singleton instance
export const categoriesApi = new CategoriesApi();
